test/test_cloudvolume.py

```diff
@@ -203,6 +210,214 @@ def test_provenance():
 
     assert cv.provenance.sources == [ 'cooldude24@princeton.edu' ]
 
+def test_info_provenance_cache():
+    image = np.zeros(shape=(128,128,128,1), dtype=np.uint8)
+    vol = create_volume_from_image(
+        image=image, 
+        offset=(0,0,0), 
+        layer_path='gs://seunglab-test/cloudvolume/caching', 
+        layer_type='image', 
+        resolution=(1,1,1), 
+        encoding='raw'
+    )
+
+    # Test Info
+    vol.cache = True
+    vol.flush_cache()
+    info = vol.refresh_info()
+    assert info is not None
+
+    with open(os.path.join(vol.cache_path, 'info'), 'r') as infof:
+        info = infof.read()
+        info = json.loads(info)
+
+    with open(os.path.join(vol.cache_path, 'info'), 'w') as infof:
+        infof.write(json.dumps({ 'wow': 'amaze' }))
+
+    info = vol.refresh_info()
+    assert info == { 'wow': 'amaze' }
+    vol.cache = False
+    info = vol.refresh_info()
+    assert info != { 'wow': 'amaze' }
+
+    # Test Provenance
+    vol.cache = True
+    vol.flush_cache()
+    prov = vol.refresh_provenance()
+    assert prov is not None
+
+    with open(os.path.join(vol.cache_path, 'provenance'), 'r') as provf:
+        prov = provf.read()
+        prov = json.loads(prov)
+
+    with open(os.path.join(vol.cache_path, 'provenance'), 'w') as provf:
+        prov['description'] = 'wow'
+        provf.write(json.dumps(prov))
+
+    prov = vol.refresh_provenance()
+    assert prov['description'] == 'wow'
+    vol.cache = False
+    prov = vol.refresh_provenance()
+    assert prov['description'] == ''
+
+
+def test_caching():
+    image = np.zeros(shape=(128,128,128,1), dtype=np.uint8)
+    image[0:64,0:64,0:64] = 1
+    image[64:128,0:64,0:64] = 2
+    image[0:64,64:128,0:64] = 3
+    image[0:64,0:64,64:128] = 4
+    image[64:128,64:128,0:64] = 5
+    image[64:128,0:64,64:128] = 6
+    image[0:64,64:128,64:128] = 7
+    image[64:128,64:128,64:128] = 8
+
+    dirpath = '/tmp/cloudvolume/caching-volume-' + str(TEST_NUMBER)
+    layer_path = 'file://' + dirpath
+
+    vol = create_volume_from_image(
+        image=image, 
+        offset=(0,0,0), 
+        layer_path=layer_path, 
+        layer_type='image', 
+        resolution=(1,1,1), 
+        encoding='raw'
+    )
+
+    vol.cache = True
+    vol.flush_cache()
+
+    # Test that reading populates the cache
+    read1 = vol[:,:,:]
+    assert np.all(read1 == image)
+
+    read2 = vol[:,:,:]
+    assert np.all(read2 == image)
+
+    assert len(os.listdir(os.path.join(vol.cache_path, vol.key))) > 0
+
+    files = os.listdir(os.path.join(vol.cache_path, vol.key))
+    validation_set = [
+        '0-64_0-64_0-64',
+        '64-128_0-64_0-64',
+        '0-64_64-128_0-64',
+        '0-64_0-64_64-128',
+        '64-128_64-128_0-64',
+        '64-128_0-64_64-128',
+        '0-64_64-128_64-128',
+        '64-128_64-128_64-128'
+    ]
+    assert set([ os.path.splitext(fname)[0] for fname in files ]) == set(validation_set)
```

Comment:
This line validates that the contents of the cache superficially match what we expect.